projects/drivers/network/network.c

```diff
@@ -0,0 +1,502 @@
+		} else {
+			/* abort? output diagnostic? */
+		}
+	} else {
+		/* abort? output diagnostic? */
+	}
+}
+
+/***************************************************************************//**
+ * @brief network_accept
+*******************************************************************************/
+err_t network_accept(void *arg, struct tcp_pcb *newpcb, err_t err)
+{
+	err_t ret_err;
+	struct network_instance *es;
+	static int32_t inst_id = 0;
+	inst_id++;
+
+	LWIP_UNUSED_ARG(arg);
+	LWIP_UNUSED_ARG(err);
+
+	/* commonly observed practive to call tcp_setprio(), why? */
+	//tcp_setprio(newpcb, TCP_PRIO_MIN);
+	tcp_setprio(newpcb, TCP_PRIO_MAX);
+
+	es = (struct network_instance *)mem_malloc(sizeof(struct network_instance));
+	if (es != NULL) {
+		es->state = ES_ACCEPTED;
+		es->pcb = newpcb;
+		es->retries = 0;
+		es->p = NULL;
+		es->instance_id = inst_id;
+		HASH_ADD_INT( instances, instance_id, es );
+#ifdef DEBUG_NETWORK
+		printf("new clientconnected: %d\n", inst_id);
+#endif // DEBUG_NETWORK
+		/* pass newly allocated es to our callbacks */
+		tcp_arg(newpcb, es);
+		tcp_recv(newpcb, network_recv);
+		tcp_err(newpcb, network_error);
+		ret_err = ERR_OK;
+	} else {
+		ret_err = ERR_MEM;
+	}
+	return ret_err;
+}
+
+/***************************************************************************//**
+ * @brief network_recv
+*******************************************************************************/
+err_t network_recv(void *arg, struct tcp_pcb *tpcb, struct pbuf *p, err_t err)
+{
+	struct network_instance *es;
+	err_t ret_err;
+	LWIP_ASSERT("arg != NULL",arg != NULL);
+	es = (struct network_instance *)arg;
+	if (p == NULL) {
+		/* remote host closed connection */
+		es->state = ES_CLOSING;
+		if(es->p == NULL) {
+			/* we're done sending, close it */
+			network_close(tpcb, es);
+		} else {
+			/* we're not done yet */
+			network_store(tpcb, es);
+		}
+		ret_err = ERR_OK;
+	} else if(err != ERR_OK) {
+		/* cleanup, for unkown reason */
+		if (p != NULL) {
+			es->p = NULL;
+			pbuf_free(p);
+		}
+		ret_err = err;
+	} else if(es->state == ES_ACCEPTED) {
+		/* first data chunk in p->payload */
+		es->state = ES_RECEIVED;
+		/* store reference to incoming pbuf (chain) */
+		es->p = p;
+		/* install send completion notifier */
+		network_store(tpcb, es);
+		ret_err = ERR_OK;
+	} else if (es->state == ES_RECEIVED) {
+		/* read some more data */
+		if(es->p == NULL) {
+			es->p = p;
+			network_store(tpcb, es);
+		} else {
+			struct pbuf *ptr;
+
+			/* chain pbufs to the end of what we recv'ed previously  */
+			ptr = es->p;
+			pbuf_chain(ptr,p);
+		}
+		ret_err = ERR_OK;
+	} else if(es->state == ES_CLOSING) {
+		/* odd case, remote side closing twice, trash data */
+		tcp_recved(tpcb, p->tot_len);
+		es->p = NULL;
+		pbuf_free(p);
+		network_close(tpcb, es);
+		ret_err = ERR_OK;
+	} else {
+		/* unkown es->state, trash data  */
+		tcp_recved(tpcb, p->tot_len);
+		es->p = NULL;
+		pbuf_free(p);
+		ret_err = ERR_OK;
+	}
+	return ret_err;
```

Comment:
Usually there is a blank line before return to separate from the rest of the code.